velox/row/UnsafeRowFast.cpp

```diff
@@ -27,12 +27,187 @@ int32_t alignBits(int32_t numBits) {
 int32_t alignBytes(int32_t numBytes) {
   return bits::roundUp(numBytes, 8);
 }
+
+// Origins from BigInteger#firstNonzeroIntNum.
+int32_t firstNonzeroIntNum(const std::vector<int32_t>& mag, int32_t length) {
+  int32_t i;
+  for (i = length - 1; i >= 0 && mag[i] == 0; --i) {
+    ;
+  }
+  return length - i - 1;
+}
+
+// Origins from BigInteger#getInt.
+int32_t getInt(
+    int32_t n,
+    int32_t sig,
+    const std::vector<int32_t>& mag,
+    int32_t length) {
+  if (n < 0) {
+    return 0;
+  }
+  if (n >= length) {
+    return sig < 0 ? -1 : 0;
+  }
+
+  int32_t magInt = mag[length - n - 1];
+  return (
+      sig >= 0 ? magInt
+               : (n <= firstNonzeroIntNum(mag, length) ? -magInt : ~magInt));
+}
+
+int32_t getBitCount(uint32_t i) {
+  static constexpr int kMaxBits = std::numeric_limits<uint64_t>::digits;
+  uint64_t num = static_cast<uint32_t>(i);
+  return bits::countBits(reinterpret_cast<uint64_t*>(&num), 0, kMaxBits);
+}
+
+// Origins from java side BigInteger#bitLength.
+int32_t
+getBitLength(int32_t sig, const std::vector<int32_t>& mag, int32_t len) {
+  int32_t n = -1;
+  if (len == 0) {
+    n = 0;
+  } else {
+    // Calculate the bit length of the magnitude.
+    int32_t magBitLength = ((len - 1) << 5) + 64 -
+        bits::countLeadingZeros(static_cast<uint64_t>(mag[0]));
+    if (sig < 0) {
+      // Check if magnitude is a power of two.
+      bool pow2 = (getBitCount((uint32_t)mag[0]) == 1);
+      for (int i = 1; i < len && pow2; ++i) {
+        pow2 = (mag[i] == 0);
+      }
+      n = (pow2 ? magBitLength - 1 : magBitLength);
+    } else {
+      n = magBitLength;
+    }
+  }
+  return n;
+}
+
+// Refer to BigInteger(byte[] val) to get the mag definition
+std::vector<uint32_t>
+convertMagArray(int64_t newHigh, uint64_t newLow, int32_t* size) {
+  std::vector<uint32_t> mag;
+  int64_t orignalLow = newLow;
+  int64_t orignalHigh = newHigh;
+  mag.push_back(newHigh >>= 32);
+  mag.push_back((uint32_t)orignalHigh);
+  mag.push_back(newLow >>= 32);
+  mag.push_back((uint32_t)orignalLow);
+
+  int32_t start = 0;
+  // Remove the front 0.
+  for (int32_t i = 0; i < 4; ++i) {
+    if (mag[i] == 0) {
+      start++;
+    }
+    if (mag[i] != 0) {
+      break;
+    }
+  }
+
+  int32_t length = 4 - start;
+  std::vector<uint32_t> newMag;
```

Comment:
Is this vector just a subset of 'mag' vector? Do we need to make a copy or can we just return a folly::Range over the necessary elements?